### PR TITLE
Rename cacheTime to gcTime

### LINKED ISSUE #4678: rename cacheTime
### context

Almost everyone gets `cacheTime` wrong ([exhibit A](https://twitter.com/TkDodo/status/1575893647392145408)). It sounds like "the amount of time that data is cached for", but that is not correct.

`cacheTime` does nothing as long as a query is still in used. It only kicks in as soon as the query becomes unused. After the time has passed, data will be "garbage collected" to avoid the cache from growing (see also [this explanation](https://twitter.com/TkDodo/status/1576078947887915013)).

RTK Query has the same feature - their prop is called [keepUnusedDataFor](https://redux-toolkit.js.org/rtk-query/api/createApi#keepunuseddatafor). I think this is quite descriptive but also a bit long.

### proposal

- rename `cacheTime` to `gcTime`

`gc` is referring to "garbage collect" time. It's a bit more technical, but also a quite [well known abbreviation](<https://en.wikipedia.org/wiki/Garbage_collection_(computer_science)>) in computer science.

Also, it is not something that most people will have to customize. The default of 5 minutes is usually fine. A rename will reduce the chance that this property is mixed up with `staleTime`.

Lastly, if someone doesn't immediately know what `gcTime` stands for, they will (hopefully) consult the docs. This is a lot better than _thinking_ they know what `cacheTime` does.

Here is an old discussion on that topic:

- https://github.com/TanStack/query/discussions/1217

### PR DESCRIPTION
**Pull request by Mamoanwar97**
closes #4678

### CODE REVIEW

File: packages/query-core/src/removable.ts [hunk 1]
```diff
@@ -11,18 +11,18 @@ export abstract class Removable {
...
     }
   }
 
-  protected updateCacheTime(newCacheTime: number | undefined): void {
-    // Default to 5 minutes (Infinity for server-side) if no cache time is set
-    this.cacheTime = Math.max(
-      this.cacheTime || 0,
-      newCacheTime ?? (isServer ? Infinity : 5 * 60 * 1000),
+  protected updateGcTime(newGcTime: number | undefined): void {
+    // Default to 5 minutes (Infinity for server-side) if no garbage collector time is set
```
Comment (TkDodo) [inline packages/query-core/src/removable.ts:22]:
```suggestion
    // Default to 5 minutes (Infinity for server-side) if no gcTime is set
```

File: packages/query-core/src/tests/hydration.test.tsx [hunk 2]
```diff
@@ -119,13 +119,13 @@ describe('dehydration and rehydration', () => {
     queryClient.clear()
   })
 
-  test('should use the cache time from the client', async () => {
+  test('should use the garbage collector time from the client', async () => {
```
Comment (TkDodo) [inline packages/query-core/src/tests/hydration.test.tsx:122]:
```suggestion
  test('should use the garbage collection time from the client', async () => {
```

File: packages/query-core/src/tests/query.test.tsx [hunk 3]
```diff
@@ -28,25 +28,25 @@ describe('query', () => {
     queryClient.clear()
   })
 
-  test('should use the longest cache time it has seen', async () => {
+  test('should use the longest garbage collector time it has seen', async () => {
```
Comment (TkDodo) [inline packages/query-core/src/tests/query.test.tsx:31]:
```suggestion
  test('should use the longest garbage collection time it has seen', async () => {
```

File: packages/query-core/src/tests/queryClient.test.tsx [hunk 4]
```diff
@@ -578,31 +578,31 @@ describe('queryClient', () => {
       expect(second).toBe(first)
     })
 
-    test('should be able to fetch when cache time is set to 0 and then be removed', async () => {
+    test('should be able to fetch when garbage collector time is set to 0 and then be removed', async () => {
```
Comment (TkDodo) [inline packages/query-core/src/tests/queryClient.test.tsx:581]:
```suggestion
    test('should be able to fetch when garbage collection time is set to 0 and then be removed', async () => {
```

File: packages/query-core/src/tests/queryClient.test.tsx [hunk 5]
```diff
@@ -578,31 +578,31 @@ describe('queryClient', () => {
...
+        gcTime: 0,
       })
       expect(result).toEqual(1)
       await waitFor(() =>
         expect(queryClient.getQueryData(key1)).toEqual(undefined),
       )
     })
 
-    test('should keep a query in cache if cache time is Infinity', async () => {
+    test('should keep a query in cache if garbage collector time is Infinity', async () => {
```
Comment (TkDodo) [inline packages/query-core/src/tests/queryClient.test.tsx:597]:
```suggestion
    test('should keep a query in cache if garbage collection time is Infinity', async () => {
```

File: packages/solid-query/src/__tests__/createQuery.test.tsx [hunk 6]
```diff
@@ -3983,7 +3983,7 @@ describe('createQuery', () => {
 
     const query = queryCache.find({ queryKey: key })
     // @ts-expect-error
-    expect(query!.cacheTimeout).toBe(undefined)
+    expect(query!.gcTimeout).toBe(undefined)
```
Comment (TkDodo) [inline packages/solid-query/src/__tests__/createQuery.test.tsx:3986]: same here, this doesn't actually do anything.

**Review by TkDodo — COMMENTED** [review]
thank you ❤️ . Just some minor comments from my side:

File: docs/react/guides/caching.md [hunk 7]
```diff
@@ -26,7 +26,7 @@ Let's assume we are using the default `cacheTime` of **5 minutes** and the defau
     - Note that regardless of whether both `fetchTodos` query functions are identical or not, both queries' [`status`](../reference/useQuery) are updated (including `isFetching`, `isLoading`, and other related values) because they have the same query key.
   - When the request completes successfully, the cache's data under the `['todos']` key is updated with the new data, and both instances are updated with the new data.
 - Both instances of the `useQuery({ queryKey: ['todos'], queryFn: fetchTodos })` query are unmounted and no longer in use.
-  - Since there are no more active instances of this query, a cache timeout is set using `cacheTime` to delete and garbage collect the query (defaults to **5 minutes**).
+  - Since there are no more active instances of this query, a cache timeout is set using `gcTime` to delete and garbage collect the query (defaults to **5 minutes**).
```
[the remark this answers is not in the document]
Comment (Mamoanwar97) [reply docs/react/guides/caching.md]: Done ✅

File: docs/react/reference/useQuery.md [hunk 8]
```diff
@@ -91,9 +91,9 @@ const {
   - Defaults to `0`
   - The time in milliseconds after data is considered stale. This value only applies to the hook it is defined on.
   - If set to `Infinity`, the data will never be considered stale
-- `cacheTime: number | Infinity`
+- `gcTime: number | Infinity`
   - Defaults to `5 * 60 * 1000` (5 minutes) or `Infinity` during SSR
-  - The time in milliseconds that unused/inactive cache data remains in memory. When a query's cache becomes unused or inactive, that cache data will be garbage collected after this duration. When different cache times are specified, the longest one will be used.
+  - The time in milliseconds that unused/inactive cache data remains in memory. When a query's cache becomes unused or inactive, that cache data will be garbage collected after this duration. When different garbage collector times are specified, the longest one will be used.
```
[the remark this answers is not in the document]
Comment (Mamoanwar97) [reply docs/react/reference/useQuery.md]: Done ✅

File: packages/query-core/src/removable.ts (hunk 1 above)
Comment (Mamoanwar97) [reply packages/query-core/src/removable.ts]: Done ✅

File: packages/query-core/src/tests/queryClient.test.tsx (hunk 4 above)
Comment (Mamoanwar97) [reply packages/query-core/src/tests/queryClient.test.tsx]: Done ✅

File: packages/query-core/src/tests/queryClient.test.tsx (hunk 5 above)
Comment (Mamoanwar97) [reply packages/query-core/src/tests/queryClient.test.tsx]: Done ✅

File: packages/react-query/src/__tests__/useQuery.test.tsx [hunk 9]
```diff
@@ -4029,7 +4029,7 @@ describe('useQuery', () => {
 
     const query = queryCache.find({ queryKey: key })
     // @ts-expect-error
-    expect(query!.cacheTimeout).toBe(undefined)
+    expect(query!.gcTimeout).toBe(undefined)
```
[the remark this answers is not in the document]
Comment (Mamoanwar97) [reply packages/react-query/src/__tests__/useQuery.test.tsx]: I tried spying on `setTimeout` here and in solidQuery tests and the tests passed. 

but for some reason in react-query tests,  jest is raising a warning and recommends using `jest.useFakeTimers()` when spying on `setTimeout`, but in solid query tests, which have the same implementation jest is not raising any warning.

when I tried to add `jest.useFakeTimers()` to the test in react-query, the test failed and useFakerTimers affected other tests to fail as well.

File: packages/solid-query/src/__tests__/createQuery.test.tsx (hunk 6 above)
Comment (Mamoanwar97) [reply packages/solid-query/src/__tests__/createQuery.test.tsx]: solved ✅

**Pull request merged**